core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -0,0 +1,118 @@
+package com.sequenceiq.cloudbreak.service.multiaz;
+
+import static com.sequenceiq.cloudbreak.common.network.NetworkConstants.SUBNET_IDS;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Service;
+
+import com.google.common.base.Strings;
+import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+import com.sequenceiq.cloudbreak.common.json.Json;
+import com.sequenceiq.cloudbreak.controller.validation.network.MultiAzValidator;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceMetaData;
+import com.sequenceiq.cloudbreak.domain.stack.instance.network.InstanceGroupNetwork;
+import com.sequenceiq.environment.api.v1.environment.model.response.DetailedEnvironmentResponse;
+
+@Service
+public class MultiAzCalculatorService {
+
+    @Inject
+    private MultiAzValidator multiAzValidator;
+
+    public Map<String, String> prepareSubnetAzMap(DetailedEnvironmentResponse environment) {
+        Map<String, String> subnetAzPairs = new HashMap<>();
+        if (environment != null && environment.getNetwork() != null) {
+            for (Map.Entry<String, CloudSubnet> entry : environment.getNetwork().getSubnetMetas().entrySet()) {
+                CloudSubnet value = entry.getValue();
+                if (!Strings.isNullOrEmpty(value.getName())) {
+                    subnetAzPairs.put(value.getName(), value.getAvailabilityZone());
+                }
+                if (!Strings.isNullOrEmpty(value.getId())) {
+                    subnetAzPairs.put(value.getId(), value.getAvailabilityZone());
+                }
+            }
+        }
+        return subnetAzPairs;
+    }
+
+    public void calculateByRoundRobin(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup) {
+        Map<String, Integer> subnetUsage = new HashMap<>();
+        Set<String> subnetIds = collectSubnetIds(instanceGroup);
+        initiateSubnetUsage(subnetAzPairs, subnetIds, subnetUsage);
+        collectCurrentSubnetUsage(instanceGroup, subnetUsage);
+
+        if (!subnetIds.isEmpty() && multiAzValidator.supportedForInstanceMetadataGeneration(instanceGroup)) {
+            for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
+                if (Strings.isNullOrEmpty(instanceMetaData.getSubnetId())) {
+                    Integer numberOfInstanceInASubnet = searchingTheLessInstanceCountForSubnets(subnetUsage);
+                    String lessUsedSubnetId = searchingTheLessUsedSubnetID(subnetUsage, numberOfInstanceInASubnet);
```

Comment:
Note: While this works perfectly, a nicer solution would be to utilize a sorted data structure. Since maps can only be sorted by their keys, one option would be to construct a `SortedSet` out of `subnetUsage.entrySet()` using a `Comparator` that compares based on the entry values (i.e. counts). This `SortedSet` would have be re-created in every loop cycle since set elements are assumed to be immutable, so any updates to the `subnetUsage` entry values would not automatically result in an updated order of existing `SortedSet` instances.

Anyway, I am happy to leave this as is if you decide so. 😄